### PR TITLE
Fix setting hasBailout when there are inlined functions in try/catch

### PR DESCRIPTION
**Pull request by meg-gupta**
Fixes OS#15078638

When we bailout executing trycode from within OP_TryCatch, we complete the execution of the current function enclosing the try/catch in the interpreter.
If there was an exception within the try region, it is caught and handled accordingly in ProcessTryHandlerBailOut which reconstructs try/catch/finally frames
when we bailout midway executing code within OP_TryCatch. If there was an exception outside the try region, the catch of the OP_TryCatch ends up catching it,
because it happens to be on the callstack. For this we use the hasBailOut bit which is per function, so we know that this exception has to be passed above.

When we have inlined functions inside the try, and for bailouts inside the inlined code, we do not set the hasBailedOut bit, so that the enclosing functions catch in OP_TryCatch catches it.

But when we bailout from inlined code inside try, we execute inlined code, as well as the enclosing function's code in the interpreter.
We will be execution code past the try/catch of the current function in the interpreter. Now if any code outside the eh region throws,
we will catch that in OP_TryCatch which happens to be on the callstack. And we will end up handling it instead of passing above because we have not set the hasBailedOutBit from the bailout point.

This change fixes this issue. We pass a pointer to the hasBailedOutBit and set it once we have finished executing the inlined frames and are ready to process the interpreter frame of the current function.

### CODE REVIEW

**Comment by meg-gupta** [comment]
I tried it out - looks much better.

**Comment by meg-gupta** [comment]
FYI, both approaches will not work as is once we start inlining functions within try/catch/finally.

**Comment by rajatd** [comment]
Hmm, yes, unfortunately

**Comment by meg-gupta** [comment]
@Penguinwizzard @agarwal-sandeep Can you help review this PR ?

File: lib/Runtime/Base/ThreadContext.h [hunk 1]
```diff
@@ -636,6 +636,7 @@ class ThreadContext sealed :
     StackProber * stackProber;
     bool isThreadBound;
     bool hasThrownPendingException;
+    bool * hasBailedOutBitOffset;
```
Comment (Penguinwizzard) [inline lib/Runtime/Base/ThreadContext.h]: Nit: This naming is slightly off, I feel, as this isn't really an offset, but is a full-fledged pointer. This becomes more of an issue in cases like https://github.com/Microsoft/ChakraCore/pull/4535/files#diff-b1ac070605da21777f9c4d31b69a2014R162, where you're setting it with a pointer plus an _actual_ offset.

File: lib/Runtime/Language/JavascriptExceptionOperators.h [hunk 2]
```diff
@@ -47,6 +47,7 @@ namespace Js
         {
           private:
             void * m_prevTryCatchFrameAddr;
+            bool m_prevDoneBailOutEhRegion;
```
Comment (Penguinwizzard) [inline lib/Runtime/Language/JavascriptExceptionOperators.h]: Nit: is this used in the current version of the changes?

Comment (meg-gupta) [reply lib/Runtime/Language/JavascriptExceptionOperators.h]: No, removed it. Thanks for catching this.

File: lib/Runtime/Base/ThreadContext.h (hunk 1 above)
Comment (meg-gupta) [reply lib/Runtime/Base/ThreadContext.h]: Done.

**Pull request merged**